source/common/buffer/buffer_impl.cc

```diff
@@ -338,55 +299,103 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
+      size = bytes_remaining;
     }
-    iovecs[num_slices_used] = slice.reserve(reservation_size);
-    bytes_remaining -= iovecs[num_slices_used].len_;
-    num_slices_used++;
-    slice_index++;
+    Slice slice(size);
+    reservation_slices.push_back(slice.reserve(size));
+    reservation_owned_slices.emplace_back(std::make_unique<SliceDataImpl>(std::move(slice)));
+    bytes_remaining -= std::min(reservation_slices.back().len_, bytes_remaining);
+  }
+
+  reservation.bufferImplUseOnlySetLength(length);
+
+  ASSERT(reservation_slices.size() <= reservation.NUM_ELEMENTS_,
+         "The absl::InlineVector should be sized so that out-of-line storage isn't needed.");
+  ASSERT(reservation_slices.size() == reservation_owned_slices.size());
+  return reservation;
+}
+
+ReservationSingleSlice OwnedImpl::reserveSingleSlice(uint64_t length, bool separate_slice) {
+  ReservationSingleSlice reservation = ReservationSingleSlice::bufferImplUseOnlyConstruct(*this);
+  if (length == 0) {
+    return reservation;
+  }
+
+  // Remove any empty slices at the end.
+  while (!slices_.empty() && slices_.back().dataSize() == 0) {
+    slices_.pop_back();
   }
 
-  // If needed, allocate one more slice at the end to provide the remainder of the reservation.
-  if (bytes_remaining != 0) {
-    slices_.emplace_back(bytes_remaining);
-    iovecs[num_slices_used] = slices_.back().reserve(bytes_remaining);
-    bytes_remaining -= iovecs[num_slices_used].len_;
-    num_slices_used++;
+  auto& reservation_slice = reservation.bufferImplUseOnlySlice();
+  auto& reservation_owned_slice = reservation.bufferImplUseOnlyOwnedSlice();
+
+  // Check whether there are any empty slices with reservable space at the end of the buffer.
+  uint64_t reservable_size =
+      (separate_slice || slices_.empty()) ? 0 : slices_.back().reservableSize();
+  if (reservable_size >= length) {
+    reservation_slice = slices_.back().reserve(length);
+    reservation_owned_slice = nullptr;
+  } else {
+    Slice slice(length);
+    reservation_slice = slice.reserve(length);
+    reservation_owned_slice = std::make_unique<SliceDataImpl>(std::move(slice));
   }
 
-  ASSERT(num_slices_used <= num_iovecs);
-  ASSERT(bytes_remaining == 0);
-  return num_slices_used;
+  return reservation;
+}
+
+void OwnedImpl::commit(uint64_t length, absl::Span<RawSlice> slices,
+                       absl::Span<SliceDataPtr> owned_slices) {
+  ASSERT(slices.size() == owned_slices.size());
+  uint64_t bytes_remaining = length;
+  for (uint32_t i = 0; i < slices.size(); i++) {
+    ASSERT((owned_slices[i] != nullptr) ==
+           (dynamic_cast<SliceDataImpl*>(owned_slices[i].get()) != nullptr));
+    std::unique_ptr<SliceDataImpl> owned_slice(
+        static_cast<SliceDataImpl*>(owned_slices[i].release()));
+
+    if (bytes_remaining > 0) {
+      if (owned_slice != nullptr) {
+        slices_.emplace_back(std::move(owned_slice->slice_));
+      }
```

Comment:
I was going to ask "Would it be possible to verify that slices[i].mem_ is within the Slice?", but that's already in place by ASSERTing that Slice::commit succeeds.